### PR TITLE
Use `softmax` rather than `argmax` and provide unit tests

### PR DESCRIPTION
**Pull request by joewallwork**
As @ma595 suggested, use of `argmax` was making the loss function non-differentiable and `softmax` should be used instead.

This PR makes this change and adds some unit tests, including one that checks the loss function values sum to 1, as should be the case for a classification problem.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -45,7 +45,8 @@ extend-select = ["D417"]
 
 # Ignore SIM108 (use ternary instead of if-else) as I think it can obscure intent.
 # Ignore RUF002 (ambiguous characters) as it does not allow apostrophes in strings.
-ignore = ["SIM108", "RUF002"]
+# Ignore S101 (use of `assert` detected) as assertions are used in the test suite.
+ignore = ["SIM108", "RUF002", "S101"]
```
Comment (jatkinson1000) [inline pyproject.toml]: You can have assert disabled only in tests, but applied to the operational code with something like:

```suggestion
ignore = ["SIM108", "RUF002"]

[tool.ruff.lint.per-file-ignores]
# Ignore S101 (use of assert) in tests
"**/test/*" = ["S101"]
```

**Review by jatkinson1000 — APPROVED** [review]
Looks sensible to me @jwallwork23
Added one suggestion that allows you to ignore asserts specifically in tests rather than the whole codebase.
